File: sdk/csharp/src/Providers/Linshiyou.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Net;
using System.Text.RegularExpressions;

namespace XxxXTeam.TempMail.Providers;

/// <summary>
/// linshiyou.com 渠道。
/// 创建邮箱：GET /api/user?user 拿 NEXUS_TOKEN cookie 与响应体邮箱，token 存 Cookie 串。
/// 收信：GET /api/mail?unseen=1 返回自定义分隔符切分的 HTML 分片，解析元信息与 srcdoc 正文。
/// </summary>
public static class Linshiyou
{
    private const string Channel = "linshiyou";
    private const string Origin = "https://linshiyou.com";

    private static readonly Dictionary<string, string> DefaultHeaders = new()
    {
        ["User-Agent"] = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0",
        ["accept-language"] = "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
        ["cache-control"] = "no-cache",
        ["dnt"] = "1",
        ["pragma"] = "no-cache",
        ["priority"] = "u=1, i",
        ["referer"] = $"{Origin}/",
        ["sec-ch-ua"] = "\"Chromium\";v=\"146\", \"Not-A.Brand\";v=\"24\", \"Microsoft Edge\";v=\"146\"",
        ["sec-ch-ua-mobile"] = "?0",
        ["sec-ch-ua-platform"] = "\"Windows\"",
        ["sec-fetch-dest"] = "empty",
        ["sec-fetch-mode"] = "cors",
        ["sec-fetch-site"] = "same-origin",
    };

    private static readonly Regex ListIdRe = new("id=\"tmail-email-list-([a-f0-9]+)\"", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex DivRe = new("<div class=\"([^\"]+)\"[^>]*>([\\s\\S]*?)</div>", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex SrcdocRe = new("\\ssrcdoc=\"([^\"]*)\"", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex DownloadRe = new("href=\"(/api/download\\?id=[^\"]+)\"", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex TagsRe = new("<[^>]+>", RegexOptions.Compiled);

    private static string StripTags(string s)
    {
        var t = TagsRe.Replace(s, " ");
        return string.Join(" ", WebUtility.HtmlDecode(t).Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries));
    }

    private static string PickDiv(string fragment, string className)
    {
        foreach (Match m in DivRe.Matches(fragment))
            if (m.Groups[1].Value == className)
                return StripTags(m.Groups[2].Value).Trim();
        return "";
    }

    private static string ParseDate(string s)
    {
        s = (s ?? "").Trim();
        if (s.Length == 0) return "";
        foreach (var fmt in new[] { "yyyy-MM-dd HH:mm:ss", "yyyy-MM-ddTHH:mm:ss" })
        {
            if (DateTime.TryParseExact(s, fmt, CultureInfo.InvariantCulture, DateTimeStyles.None, out var dt))
                return new DateTimeOffset(dt, TimeSpan.Zero).ToString("yyyy-MM-ddTHH:mm:ssZ", CultureInfo.InvariantCulture);
        }
        return "";
    }

    private static string ExtractSrcdoc(string bodyPart)
    {
        var m = SrcdocRe.Match(bodyPart);
        return m.Success ? WebUtility.HtmlDecode(m.Groups[1].Value) : "";
    }

    private static List<Email> ParseSegments(string raw, string recipient)
    {
        raw = (raw ?? "").Trim();
        var outList = new List<Email>();
        if (raw.Length == 0) return outList;

        foreach (var segRaw in raw.Split("<-----TMAILNEXTMAIL----->"))
        {
            var seg = segRaw.Trim();
            if (seg.Length == 0) continue;
            var parts = seg.Split("<-----TMAILCHOPPER----->", 2);
            var listPart = parts[0];
            var bodyPart = parts.Length > 1 ? parts[1] : "";

            var mid = ListIdRe.Match(listPart);
            if (!mid.Success) continue;
            var midStr = mid.Groups[1].Value;

            var fromList = PickDiv(listPart, "name");
            var subjectList = PickDiv(listPart, "subject");
            var previewList = PickDiv(listPart, "body");

            var fromBody = PickDiv(bodyPart, "tmail-email-sender");
            var timeBody = PickDiv(bodyPart, "tmail-email-time");
            var titleBody = PickDiv(bodyPart, "tmail-email-title");
            var htmlBody = ExtractSrcdoc(bodyPart);

            var fromAddr = fromBody.Length > 0 ? fromBody : fromList;
            var subject = titleBody.Length > 0 ? titleBody : subjectList;
            var text = previewList.Length > 0 ? previewList : StripTags(htmlBody);

            var attachments = new List<EmailAttachment>();
            var dm = DownloadRe.Match(bodyPart);
            if (dm.Success)
                attachments.Add(new EmailAttachment { Filename = "", Url = $"{Origin}{dm.Groups[1].Value}" });

            outList.Add(new Email
            {
                Id = midStr,
                From = fromAddr,
                To = recipient,
                Subject = subject,
                Text = text,
                Html = htmlBody,
                Date = ParseDate(timeBody),
                IsRead = false,
                Attachments = attachments,
            });
        }
        return outList;
    }

    /// <summary>创建临时邮箱</summary>
    public static EmailInfo Generate()
    {
        var resp = Http.Get($"{Origin}/api/user?user", DefaultHeaders);
        resp.EnsureSuccess();

        var nexus = ProviderHttpUtil.ExtractCookieValue(resp, "NEXUS_TOKEN");
        if (nexus.Length == 0) throw new Exception("linshiyou: NEXUS_TOKEN not in Set-Cookie");

        var email = (resp.Body ?? "").Trim();
        if (email.Length == 0 || !email.Contains('@')) throw new Exception("linshiyou: invalid email in response body");

        var token = $"NEXUS_TOKEN={nexus}; tmail-emails={email}";
        return new EmailInfo(Channel, email, token);
    }

    /// <summary>获取邮件列表</summary>
    public static List<Email> GetEmails(string? token, string email)
    {
        var headers = new Dictionary<string, string>(DefaultHeaders)
        {
            ["Cookie"] = token ?? "",
            ["x-requested-with"] = "XMLHttpRequest",
        };
        var resp = Http.Get($"{Origin}/api/mail?unseen=1", headers);
        resp.EnsureSuccess();
        return ParseSegments(resp.Body, email);
    }
}

File: sdk/php/src/Providers/Linshiyou.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail\Providers;

use ChanhanzhanX\TempMail\Email;
use ChanhanzhanX\TempMail\EmailAttachment;
use ChanhanzhanX\TempMail\EmailInfo;
use ChanhanzhanX\TempMail\HttpClient;

/**
 * linshiyou.com 渠道实现（临时邮）
 *
 * GET /api/user?user 返回响应体即邮箱地址，Set-Cookie 含 NEXUS_TOKEN；
 * token = "NEXUS_TOKEN=...; tmail-emails=email"；
 * getEmails GET /api/mail?unseen=1 返回 HTML 分片，以自定义分隔符切分邮件。
 */
final class Linshiyou
{
    private const ORIGIN = 'https://linshiyou.com';

    private const SEG_SPLIT = '<-----TMAILNEXTMAIL----->';
    private const SEG_CHOP = '<-----TMAILCHOPPER----->';

    /** @var array<string,string> */
    private const HEADERS = [
        'User-Agent' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 '
            . '(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0',
        'accept-language' => 'zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6',
        'cache-control' => 'no-cache',
        'dnt' => '1',
        'pragma' => 'no-cache',
        'priority' => 'u=1, i',
        'referer' => self::ORIGIN . '/',
        'sec-ch-ua' => '"Chromium";v="146", "Not-A.Brand";v="24", "Microsoft Edge";v="146"',
        'sec-ch-ua-mobile' => '?0',
        'sec-ch-ua-platform' => '"Windows"',
        'sec-fetch-dest' => 'empty',
        'sec-fetch-mode' => 'cors',
        'sec-fetch-site' => 'same-origin',
    ];

    private static function stripTags(string $s): string
    {
        $t = preg_replace('/<[^>]+>/', ' ', $s) ?? '';
        $t = html_entity_decode($t, ENT_QUOTES | ENT_HTML5);
        return implode(' ', preg_split('/\s+/', trim($t)) ?: []);
    }

    /**
     * 在 HTML 分片中提取指定 class 的 div 文本（class 需精确匹配）。
     */
    private static function pickDiv(string $fragment, string $className): string
    {
        if (preg_match_all('/<div class="([^"]+)"[^>]*>([\s\S]*?)<\/div>/i', $fragment, $ms, PREG_SET_ORDER)) {
            foreach ($ms as $m) {
                if ($m[1] === $className) {
                    return trim(self::stripTags($m[2]));
                }
            }
        }
        return '';
    }

    private static function parseDate(string $s): string
    {
        $s = trim($s);
        if ($s === '') {
            return '';
        }
        $ts = strtotime($s);
        if ($ts !== false) {
            return gmdate('Y-m-d\TH:i:s\Z', $ts);
        }
        return '';
    }

    private static function extractSrcdoc(string $bodyPart): string
    {
        if (preg_match('/\ssrcdoc="([^"]*)"/i', $bodyPart, $m) === 1) {
            return html_entity_decode($m[1], ENT_QUOTES | ENT_HTML5);
        }
        return '';
    }

    /**
     * @return Email[]
     */
    private static function parseSegments(string $raw, string $recipient): array
    {
        $raw = trim($raw);
        if ($raw === '') {
            return [];
        }
        $out = [];
        foreach (explode(self::SEG_SPLIT, $raw) as $seg) {
            $seg = trim($seg);
            if ($seg === '') {
                continue;
            }
            $parts = explode(self::SEG_CHOP, $seg, 2);
            $listPart = $parts[0];
            $bodyPart = $parts[1] ?? '';

            if (preg_match('/id="tmail-email-list-([a-f0-9]+)"/i', $listPart, $mid) !== 1) {
                continue;
            }
            $midStr = $mid[1];

            $fromList = self::pickDiv($listPart, 'name');
            $subjectList = self::pickDiv($listPart, 'subject');
            $previewList = self::pickDiv($listPart, 'body');

            $fromBody = self::pickDiv($bodyPart, 'tmail-email-sender');
            $timeBody = self::pickDiv($bodyPart, 'tmail-email-time');
            $titleBody = self::pickDiv($bodyPart, 'tmail-email-title');
            $htmlBody = self::extractSrcdoc($bodyPart);

            $fromAddr = $fromBody !== '' ? $fromBody : $fromList;
            $subject = $titleBody !== '' ? $titleBody : $subjectList;
            $text = $previewList !== '' ? $previewList : self::stripTags($htmlBody);

            $attachments = [];
            if (preg_match('#href="(/api/download\?id=[^"]+)"#i', $bodyPart, $dm) === 1) {
                $attachments[] = new EmailAttachment(filename: '', url: self::ORIGIN . $dm[1]);
            }

            $out[] = new Email(
                id: $midStr,
                fromAddr: $fromAddr,
                to: $recipient,
                subject: $subject,
                text: $text,
                html: $htmlBody,
                date: self::parseDate($timeBody),
                isRead: false,
                attachments: $attachments,
            );
        }
        return $out;
    }

    public static function generate(): EmailInfo
    {
        $resp = HttpClient::get(self::ORIGIN . '/api/user?user', self::HEADERS);
        if ($resp->getStatusCode() >= 400) {
            throw new \RuntimeException('linshiyou: 请求失败 ' . $resp->getStatusCode());
        }
        $nexus = '';
        foreach ($resp->getHeader('Set-Cookie') as $sc) {
            if (preg_match('/NEXUS_TOKEN=([^;]+)/', $sc, $m) === 1) {
                $nexus = $m[1];
                break;
            }
        }
        if ($nexus === '') {
            throw new \RuntimeException('linshiyou: NEXUS_TOKEN not in Set-Cookie');
        }
        $email = trim((string) $resp->getBody());
        if ($email === '' || !str_contains($email, '@')) {
            throw new \RuntimeException('linshiyou: invalid email in response body');
        }
        $token = 'NEXUS_TOKEN=' . $nexus . '; tmail-emails=' . $email;
        return new EmailInfo('linshiyou', $email, $token);
    }

    /**
     * @return Email[]
     */
    public static function getEmails(string $email, ?string $token): array
    {
        if ($token === null || $token === '') {
            throw new \InvalidArgumentException('linshiyou: token 不能为空');
        }
        $headers = self::HEADERS;
        $headers['Cookie'] = $token;
        $headers['x-requested-with'] = 'XMLHttpRequest';
        $resp = HttpClient::get(self::ORIGIN . '/api/mail?unseen=1', $headers);
        if ($resp->getStatusCode() >= 400) {
            throw new \RuntimeException('linshiyou: 读信失败 ' . $resp->getStatusCode());
        }
        return self::parseSegments((string) $resp->getBody(), $email);
    }
}
